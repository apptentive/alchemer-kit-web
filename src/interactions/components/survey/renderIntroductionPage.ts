import renderDisclaimerText from './renderDisclaimerText';

const introductionSelector = 'apptentive-survey-introduction';

/**
 * Renders a static text step for a survey.
 *
 * @param {HTMLElement} parentContainer - The parent container the intro appears within.
 * @param {string} introText - The static introduction text to display.
 * @param {string} buttonText - The text to use for the button text.
 * @param {string} disclaimerText - The text to use for the disclaimer beneath the button.
 * @param {string} revealedElementSelector - The HTML selector to use for the element to display when the button is clicked.
 * @returns {HTMLElement} - The HTML element containing the introduction content.
 */
const renderIntroductionPage = (
  parentContainer: HTMLElement,
  introText?: string,
  buttonText?: string,
  disclaimerText?: string,
  revealedElementSelector = 'form.apptentive-survey-questions'
) => {
  const nextButtonNode = document.createElement('button');
  nextButtonNode.type = 'button';
  nextButtonNode.classList.add('apptentive-button--primary');
  nextButtonNode.textContent = buttonText || 'Start Survey';
  nextButtonNode.title = buttonText || 'Start Survey';
  nextButtonNode.addEventListener('click', (_) => {
    const introductionElement = parentContainer.querySelector<HTMLElement>(introductionSelector);
    const formElement = parentContainer.querySelector<HTMLFormElement>(revealedElementSelector);

    if (introductionElement && formElement) {
      introductionElement.style.display = 'none';
      formElement.style.display = '';
    }
  });

  const staticActionsNode = document.createElement('div');
  staticActionsNode.classList.add('apptentive-message__actions');
  staticActionsNode.append(nextButtonNode);

  // If there is disclaimer text render it here, otherwise add a class to format the intro page differently
  if (disclaimerText) {
    staticActionsNode.append(renderDisclaimerText(disclaimerText));
  } else {
    staticActionsNode.classList.add('apptentive-message__actions--tall');
  }

  const introductionNode = document.createElement(introductionSelector);
  introductionNode.classList.add('apptentive-message');

  if (introText) {
    const staticContentNode = document.createElement('p');
    staticContentNode.className = 'apptentive-message__content';
    staticContentNode.innerText = introText;

    const staticWrapperNode = document.createElement('div');
    staticWrapperNode.classList.add('apptentive-message__content-wrapper');
    staticWrapperNode.append(staticContentNode);

    introductionNode.append(staticWrapperNode);
  }

  introductionNode.append(staticActionsNode);

  return introductionNode;
};

export default renderIntroductionPage;
