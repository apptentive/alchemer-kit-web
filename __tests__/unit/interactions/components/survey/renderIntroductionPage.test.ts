import renderIntroductionPage from '../../../../../src/interactions/components/survey/renderIntroductionPage';

const _introText = 'Intro Text';
const _disclaimerText = 'Short Disclaimer';
const _buttonText = 'Button Text';

describe('renderIntroductionPage', () => {
  afterEach(() => {
    document.body.innerHTML = '';
  });

  test('properly renders the introduction', () => {
    const output = renderIntroductionPage(document.body, _introText, _buttonText, _disclaimerText);

    expect(output).toMatchSnapshot();
  });

  test('properly renders the introduction without intro text', () => {
    const output = renderIntroductionPage(document.body, undefined, _buttonText, _disclaimerText);

    expect(output).toMatchSnapshot();
  });

  test('properly renders the introduction without button text', () => {
    const output = renderIntroductionPage(document.body, _introText, undefined, _disclaimerText);

    expect(output).toMatchSnapshot();
  });

  test('properly renders the introduction without disclaimer text', () => {
    const output = renderIntroductionPage(document.body, _introText, _buttonText, undefined);

    expect(output).toMatchSnapshot();
  });

  test('properly unhides the element selector when the next button is clicked', () => {
    document.body.innerHTML = `
      <div class="parent-container">
        <form class="apptentive-survey-questions" style="display: none;">
          <p>Element to reveal</p>
        </form>
      </div>
    `;

    const parentContainer = document.body.querySelector<HTMLDivElement>('.parent-container')!;
    const introductionElement = renderIntroductionPage(parentContainer, _introText, _buttonText, _disclaimerText);
    parentContainer.append(introductionElement);

    const formElement = parentContainer.querySelector<HTMLElement>('.apptentive-survey-questions')!;
    const button = parentContainer.querySelector<HTMLButtonElement>('.apptentive-button--primary')!;
    button.click();

    expect(formElement.style.display).toBe('');
    expect(introductionElement.style.display).toBe('none');
  });

  test('properly uses the custom element selector when passed in', () => {
    document.body.innerHTML = `
      <div class="parent-container">
        <form class="custom-element-selector" style="display: none;">
          <p>Element to reveal</p>
        </form>
      </div>
    `;

    const parentContainer = document.body.querySelector<HTMLDivElement>('.parent-container')!;
    const introductionElement = renderIntroductionPage(
      parentContainer,
      _introText,
      _buttonText,
      _disclaimerText,
      '.custom-element-selector'
    );
    parentContainer.append(introductionElement);

    const formElement = parentContainer.querySelector<HTMLElement>('.custom-element-selector')!;
    const button = parentContainer.querySelector<HTMLButtonElement>('.apptentive-button--primary')!;
    button.click();

    expect(formElement.style.display).toBe('');
    expect(introductionElement.style.display).toBe('none');
  });

  test('properly handles invalid custom element selector', () => {
    document.body.innerHTML = `
      <div class="parent-container"></div>
    `;

    const parentContainer = document.body.querySelector<HTMLDivElement>('.parent-container')!;
    const introductionElement = renderIntroductionPage(
      parentContainer,
      _introText,
      _buttonText,
      _disclaimerText,
      '.invalid-element-selector'
    );
    parentContainer.append(introductionElement);

    const button = parentContainer.querySelector<HTMLButtonElement>('.apptentive-button--primary')!;
    button.click();

    expect(introductionElement.style.display).toBe('');
  });
});
