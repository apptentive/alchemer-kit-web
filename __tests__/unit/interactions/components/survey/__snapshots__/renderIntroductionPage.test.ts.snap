// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renderIntroductionPage properly renders the introduction 1`] = `
<apptentive-survey-introduction
  class="apptentive-message"
>
  <div
    class="apptentive-message__content-wrapper"
  >
    <p
      class="apptentive-message__content"
    />
  </div>
  <div
    class="apptentive-message__actions"
  >
    <button
      class="apptentive-button--primary"
      title="Button Text"
      type="button"
    >
      Button Text
    </button>
    <div
      class="apptentive-disclaimer"
    >
      <p
        class="apptentive-disclaimer__text apptentive-typography__description apptentive-typography__description--tall"
      >
        Short Disclaimer
      </p>
    </div>
  </div>
</apptentive-survey-introduction>
`;

exports[`renderIntroductionPage properly renders the introduction without button text 1`] = `
<apptentive-survey-introduction
  class="apptentive-message"
>
  <div
    class="apptentive-message__content-wrapper"
  >
    <p
      class="apptentive-message__content"
    />
  </div>
  <div
    class="apptentive-message__actions"
  >
    <button
      class="apptentive-button--primary"
      title="Start Survey"
      type="button"
    >
      Start Survey
    </button>
    <div
      class="apptentive-disclaimer"
    >
      <p
        class="apptentive-disclaimer__text apptentive-typography__description apptentive-typography__description--tall"
      >
        Short Disclaimer
      </p>
    </div>
  </div>
</apptentive-survey-introduction>
`;

exports[`renderIntroductionPage properly renders the introduction without disclaimer text 1`] = `
<apptentive-survey-introduction
  class="apptentive-message"
>
  <div
    class="apptentive-message__content-wrapper"
  >
    <p
      class="apptentive-message__content"
    />
  </div>
  <div
    class="apptentive-message__actions apptentive-message__actions--tall"
  >
    <button
      class="apptentive-button--primary"
      title="Button Text"
      type="button"
    >
      Button Text
    </button>
  </div>
</apptentive-survey-introduction>
`;

exports[`renderIntroductionPage properly renders the introduction without intro text 1`] = `
<apptentive-survey-introduction
  class="apptentive-message"
>
  <div
    class="apptentive-message__actions"
  >
    <button
      class="apptentive-button--primary"
      title="Button Text"
      type="button"
    >
      Button Text
    </button>
    <div
      class="apptentive-disclaimer"
    >
      <p
        class="apptentive-disclaimer__text apptentive-typography__description apptentive-typography__description--tall"
      >
        Short Disclaimer
      </p>
    </div>
  </div>
</apptentive-survey-introduction>
`;
